### ISSUE TITLE
wire: document why wire doesn't allow duplicate identical providers in a provider set

### ISSUE THREAD

**Issue by oceanful** [google/go-cloud#492]
`wire.NewSet()` is the perfect terminology. By that definition, this (albeit contrived example) should ideally work:

```golang
var Set = wire.NewSet(ProvideFoo, ProvideFoo)
```

Currently, this fails with a multiple bindings error. While you certainly want to prohibit the application from defining bindings in conflicting ways, automatically deduplicating _identical_ bindings would simplify application development with Wire.

A more interesting example is when you have provider sets at multiple levels of the stack with inter-dependencies:

```
                 ----------------
                /| feature1.Set |\
               / ---------------- \   ----------------
              /                    ---| library1.Set |
------------ /   ---------------- /   ----------------
| main.Set | ----| feature2.Set |/
------------ \   ----------------
              \
               \ ----------------     ----------------
                \| feature3.Set |-----| library2.Set |
                 ----------------     ----------------
```

Without deduplication, `main.Set` has to list all of the Sets in the entire application:

```golang
package main
var Set = wire.Set(
    feature1.Set,
    feature2.Set,
    feature3.Set,
    library1.Set,
    library2.Set)
```

This "top level aggregation" pattern requires knowledge of the entire tree of Sets at the top level, and can be a maintenance hazard as Sets are removed / reconfigured (i.e. dependency rot). This is a common problem with other DI frameworks like Guice.

*With* deduplication, each Set can be self-contained by including the Sets of their dependencies, even shared ones. Then the top level need only include Sets for its high level features:

```golang
package feature1
var Set = wire.Set(..., library1.Set)

package feature2
var Set = wire.Set(..., library1.Set)

package feature3
var Set = wire.Set(..., library2.Set)

package main
var MainSet = wire.Set(
    feature1.Set,
    feature2.Set,
    feature3.Set)
```

This provides better encapsulation, giving each package the sovereignty to add and remove downstream dependencies without requiring modifications to their upstream dependents.

Long story short: deduplication of identical bindings can allow applications to declare (possibly shared) provider set dependencies, improving encapsulation and making application "wiring" closer to dependency-aware build systems like Bazel.

I hope this provides some good context for this feature request. Thanks!

**Comment by vangent** [comment]
HI @oceanful , thanks for the suggestion. Sorry for the slow response, we have been discussing this internally.

**Comment by vangent** [comment]
I discussed this with @zombiezen , and we think that provider sets should be defined to avoid the issue you describe.

In your example, if `library1` is an internal implementation detail of `feature1`, then the constructors for `feature1` objects should not expose it. `feature1` should just create anything from `library1` it needs explicitly. The provider set for `feature1` shouldn't expose anything about `library1` either, so the problem you describe is avoided. Example: imagine instead of `library1` we have `priorityqueue`, with a priority queue implementation. `feature1` uses a priority queue but only as an implementation details. `feature1.New` might look something like:

```
func New() {
  return &impl{pq: priorityqueue.New(), ...}
}
```

--> The priority queue is an implementation details of `feature1` and is not exposed to the app at all.

On the other hand, if `feature1` **does** expose objects from `library1` as something to be injected, then it is no longer an internal implementation detail, and the top-level is the right place to decide what should be injected. The constructor argument should be an interface; if it's not an interface, then there's no choice about what to pass, and it probably shouldn't be exposed at all (instead, use of the previous approach of constructing it internally).

```
func New(pq priorityqueue.Interface) {
  return &impl{pq: pq, ...}
}
```

Another possibility is that `library1` has dependencies (e.g., maybe it needs a `Logger`). `feature1` should expose them to the top-level app, e.g.:

```
func New(log logger.Interface) {
  return &impl{pq: priorityQueue.New(log), ...}
}
```

Again, the top-level app is the right place to decide what to inject for the logger; any provider sets for `priorityqueue` and `feature1` should not include a provider for constructing a `logger`, the app should supply it.


@zombiezen for any additional comments or clarifications.

@oceanful , thoughts?

**Comment by oceanful** [comment]
Thank you for considering and discussing this issue! 

I think "library" was a poor choice of naming, as the use case in my mind was not implementation details like priority queues, for which each component would indeed want its own. (In fact, I don't think Wire is designed for non-singleton objects anyway).

The use case that deduplication is useful for is more for the latter case, where one or more high-level components depend on a shared service. `Logger` isn't quite the best example because one can imagine multiple implementations of it, at which point it indeed makes sense for the top level to choose the desired implementation. A better example is a service with a single implementation (let's consider a `NetService` as an example), whereby you want dependents to share a single implementation.

The nice thing about allowing each component to declare such dependencies is that they allow components to be self contained. If you install `feature1.Set`, it automatically pulls in its dependencies  rather than requiring the top level to know. `feature2.Set` would be able to depend on the same `NetService` by virtue of deduplication, and if the top level removes `feature1.Set` and `feature2.Set`, the `NetService` dependency is eliminated because it is no longer reachable from the top level (thereby avoiding the dependency rot issue).

This issue happened for us a lot when building large Java apps with Guice. Perhaps it is not so much of a problem for Golang, so you're welcome to dismiss it if you don't think it's relevant. On the other hand, since Wire is very much designed for provision of singletons, it doesn't seem like it would be harmful to perform deduplication of identical bindings, while it would allow the encapsulation described above.

Thanks again for considering this; I appreciate it!

**Comment by zombiezen** [comment]
Assuming that it is an error to have the same type provided by different providers, the proposed behavior seems to constrain [provider set evolution](https://github.com/google/go-cloud/blob/master/wire/README.md#evolving-provider-sets) even further. Right now, it is safe to (emphasis mine):

> **Change which provider a provider set uses to provide a specific output**, as long as it does not introduce a new input to the provider set. It may remove inputs. However, note that existing injectors will use the old provider until they are regenerated.

With the proposed change, this could be a breaking provider set change, since the user of the provider set might assume that two provider sets are compatible by bringing in the same function.

I recognize the problem you're pointing out, @oceanful. I don't think this is a Go versus Java, or even Wire versus Guice, problem. I see the proposal at handing adding complexity to the model with not enough gained. I'm closing this issue out because I don't see this working out for Wire at this time. I'm open to hearing different ways of trying to tackle the problem, but I don't think this is the way to do it.

**Comment by vangent** [comment]
I would like to add something about this discussion to the Wire documentation, so re-opening for that.
